dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java

```diff
@@ -0,0 +1,190 @@
+        {
+            UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+            return userDetails.getUsername();
+        }
+        else if ( principal instanceof Dhis2User )
+        {
+            Dhis2User dhisOidcUser = (Dhis2User) authentication.getPrincipal();
+            return dhisOidcUser.getUsername();
+        }
+        else
+        {
+            throw new RuntimeException( "Authentication principal is not supported; principal:" + principal );
+        }
+    }
+
+    public static User getCurrentUser()
+    {
+        String username = getCurrentUsername();
+
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        if ( authentication == null || !authentication.isAuthenticated() || authentication.getPrincipal() == null )
+        {
+            return null;
+        }
+
+        if ( username == null )
+        {
+            throw new IllegalStateException( "No current user" );
+        }
+
+        if ( username.equals( "anonymousUser" ) )
+        {
+            return null;
+        }
+
+        Object principal = authentication.getPrincipal();
+
+        if ( principal instanceof UserDetails )
+        {
+            User principal1 = (User) authentication.getPrincipal();
+            initializeUser( principal1 );
+
+            return principal1;
+        }
+        else if ( principal instanceof Dhis2User )
+        {
+            Dhis2User dhisOidcUser = (Dhis2User) authentication.getPrincipal();
+
+            return dhisOidcUser.getDhis2User();
+        }
+        else
+        {
+            throw new RuntimeException( "Authentication principal is not supported; principal:" + principal );
+        }
+    }
+
+    public static void setUserSetting( UserSettingKey key, Serializable value )
+    {
+        setUserSettingInternal( key.getName(), value );
+    }
+
+    public static void setUserSettingInternal( String key, Serializable value )
+    {
+        User currentUser = getCurrentUser();
+        if ( currentUser != null )
+        {
+            Map<String, Serializable> userSettings = currentUser.getUserSettings();
+            if ( userSettings != null )
+            {
+                if ( value != null )
+                {
+                    userSettings.put( key, value );
+                }
+                else
+                {
+                    userSettings.remove( key );
+                }
+            }
+        }
+    }
+
+    public static <T> T getUserSetting( UserSettingKey key )
+    {
+        User currentUser = getCurrentUser();
+        if ( currentUser == null )
+        {
+            return null;
+        }
+
+        Map<String, Serializable> userSettings = currentUser.getUserSettings();
+        if ( userSettings == null )
+        {
+            return null;
+        }
+
+        return (T) userSettings.get( key.getName() );
+    }
+
+    public static void initializeUser( User user )
```

Comment:
This is the big issue here, we might want to split it, but this will require a lot of refactoring. It will be easier/less changes after this refactoring I believe. We should discuss this more.